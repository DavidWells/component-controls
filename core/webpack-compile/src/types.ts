import * as webpack from 'webpack';
import { RuleTypes } from '@component-controls/webpack-configs';
import { LoadingStore } from '@component-controls/loader';

/**
 * configuration properties for compile and run
 */
export interface CompileProps {
  /**
   * webpack configuration object
   */
  webPack?: webpack.Configuration;
  /**
   * a list of webpack configuration presets from webpack-configs packages
   */
  presets?: RuleTypes;
  /**
   * path to the configuration file e.g : '.storybook'
   */
  configPath?: string;

  /** public output folder for the bundle */
  outputFolder?: string;
}

/**
 * return type from compile and watch functions
 */
export interface CompileResults {
  /**
   * the stories store
   */
  store: LoadingStore;
  /**
   * the webpack stats object
   */
  stats: webpack.Stats;
}

/**
 * adds webpack WatchOptions to the Compiler options
 */
export type WatchProps = {
  watchOptions?: webpack.ICompiler.WatchOptions;
} & CompileProps;
