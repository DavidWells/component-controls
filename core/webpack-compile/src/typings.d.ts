declare module '@component-controls/loader/plugin';
