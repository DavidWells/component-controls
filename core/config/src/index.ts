import * as path from 'path';
import * as fs from 'fs';
import { sync as globSync } from 'glob';
import yargs from 'yargs';
import { Configuration } from '@component-controls/specification';

export const configFileNames = ['main.js', 'main.json', 'main.ts'];

export interface ConfigrationResult {
  config: Configuration;
  configPath: string;
}
/**
 * return the configration object
 * command line accepts -c/ -config parameter for config path
 * the config file is assumed named main.js/main.ts
 * @param baseFolder project folder to start the searh with
 */
export const getConfiguration = (
  baseFolder: string,
  args: string[] = process.argv,
): ConfigrationResult | undefined => {
  const argv = yargs(args)
    .option('config', {
      alias: 'c',
      description: 'configuration folder',
      type: 'string',
    })
    .help()
    .alias('help', 'h').argv;

  const configPath = argv.config
    ? path.resolve(baseFolder, argv.config)
    : baseFolder;
  const configFiles = fs.readdirSync(configPath);
  const configFile = configFiles.find(file =>
    configFileNames.includes(file.toLowerCase()),
  );

  return configFile
    ? {
        config: require(path.resolve(configPath, configFile)),
        configPath,
      }
    : undefined;
};

/**
 * find the story files out of a configuration file
 * using glob for the regex file search
 * @param param0 configuration info return by getConfiguration
 */
export const extractStories = ({
  config,
  configPath,
}: ConfigrationResult): string[] | undefined => {
  const stories =
    config && config.stories
      ? config.stories.reduce((acc: string[], storyRg: string) => {
          const regex = storyRg.replace(/\((.*?)\)/g, '+($1)');
          return [...acc, ...globSync(path.resolve(configPath, regex))];
        }, [])
      : undefined;
  return stories;
};
