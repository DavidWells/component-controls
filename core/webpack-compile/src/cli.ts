import { compile, watch } from './index';
import { cliArgs, CliArgTypes, ArgOptions } from './args';
import { LogLevel } from '@component-controls/logger';

export { cliArgs, defaultCliArgs, ArgOptions } from './args';

export const getArgs = (options?: ArgOptions): CliArgTypes => {
  const argv = cliArgs(options);
  return argv.help().alias('help', 'h').argv;
};
export const run = async (options?: ArgOptions) => {
  const argv = getArgs(options);
  process.env.NODE_ENV = argv.mode;
  const exec = argv.watch ? watch : compile;
  await exec({
    configPath: argv.config,
    presets: argv.presets,
    bundleName: argv.bundle,
    distFolder: argv.dist,
    staticFolder: argv.static,
    logOptions: {
      logLevel: argv.loglevel as LogLevel,
    },
  });
};
