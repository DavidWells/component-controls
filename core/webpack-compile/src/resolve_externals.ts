import { Configuration } from 'webpack';

export interface ResolveExternalsConfig {
  resolve: { alias: { [key: string]: string } };
  externals: Configuration['externals'];
}

export interface ExternalProps {
  lib: string;
  externalName?: string | { [key: string]: string };
}

export class ResolveExternals {
  private _config: ResolveExternalsConfig;
  constructor(config: ResolveExternalsConfig) {
    this._config = config;
  }
  addAlias = (props: ExternalProps): void => {
    const { lib, externalName } = props;
    if (typeof this._config.externals !== 'object') {
      this._config.externals = {};
    }
    (this._config.externals as Record<string, unknown>)[lib] =
      externalName || lib;
  };
}
