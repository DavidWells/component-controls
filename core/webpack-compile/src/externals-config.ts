import { ExternalProps } from './resolve_externals';
export const defaultExternals: ExternalProps[] = [
  {
    lib: 'react',
    externalName: {
      commonjs: 'react',
      commonjs2: 'react',
      amd: 'React',
      root: 'React',
    },
  },
  {
    lib: 'react-dom',
    externalName: {
      commonjs: 'react-dom',
      commonjs2: 'react-dom',
      amd: 'ReactDOM',
      root: 'ReactDOM',
    },
  },
  {
    lib: 'prop-types',
    externalName: {
      commonjs: 'prop-types',
      commonjs2: 'prop-types',
      amd: 'PropTypes',
      root: 'PropTypes',
    },
  },
  { lib: 'regenerator-runtime' },
  {
    lib: 'lodash',
    externalName: {
      commonjs: 'lodash',
      commonjs2: 'lodash',
      amd: 'lodash',
      root: '_',
    },
  },

  { lib: 'theme-ui' },
  { lib: '@theme-ui/presets' },
  { lib: '@theme-ui/core' },
  { lib: '@theme-ui/css' },

  { lib: '@component-controls/core' },
  { lib: '@component-controls/store' },
  { lib: '@component-controls/blocks' },
  { lib: '@component-controls/components' },

  { lib: 'broadcast-channel' },

  { lib: 'react-helmet' },
  { lib: 'react-table' },
  { lib: 'react-colorful' },
  { lib: 'react-google-charts' },
  { lib: 'react-tabs' },
  { lib: 'react-switch' },
  { lib: 'react-popper' },
  { lib: 'react-popper-tooltip' },
  { lib: 'react-animate-height' },
  { lib: 'polished' },
  { lib: 'tinycolor2' },

  { lib: 'prism-react-renderer/themes' },
  { lib: 'prism-react-renderer' },

  { lib: '@primer/octicons-react' },
  { lib: '@mdx-js/react' },

  { lib: 'markdown-to-jsx' },

  { lib: 'axe-core' },
  { lib: 'faker/locale/en_US' },
  { lib: 'fuse.js' },
];
