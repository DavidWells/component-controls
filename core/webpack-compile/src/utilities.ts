import webpack, {
  Configuration,
  Compiler,
  Stats,
  HotModuleReplacementPlugin,
} from 'webpack';
import * as path from 'path';
const chalk = require('chalk');
import LoaderPlugin from '@component-controls/loader/plugin';
import {
  mergeWebpackConfig,
  deepMergeWebpackConfig,
} from '@component-controls/webpack-configs';
import { loadConfiguration } from '@component-controls/config';
import { CompileProps, CompileResults } from './types';
import { ResolveExternals, ResolveExternalsConfig } from './resolve_externals';
import { defaultExternals } from './externals-config';

export type CompileRunProps = CompileProps & {
  /**
   * webpack mode
   */
  mode: Configuration['mode'];
};

const createConfig = (options: CompileRunProps): webpack.Configuration => {
  const { webPack, presets, configPath, mode, outputFolder } = options;
  const distFolder = path.resolve(__dirname, '../dist');
  const plugins = [
    new LoaderPlugin({
      config: configPath,
      escapeOutput: mode === 'development',
    }),
    mode === 'development' && new HotModuleReplacementPlugin({}),
  ].filter(Boolean);
  const webpackConfig = mergeWebpackConfig(
    {
      mode,
      entry: {
        stories: require.resolve(
          '@component-controls/loader/story-store-data.js',
        ),
      },
      output: {
        path: distFolder,
        filename: 'bundle.js',
        libraryTarget: 'umd',
        globalObject: 'this',
      },
      resolve: {
        alias: {},
      },
      externals: {},
      plugins,
      ...(webPack || {}),
    },
    presets,
    { outputFolder, distFolder },
  );

  //add all the aliases to avoid double loading of packages
  const alias = new ResolveExternals(webpackConfig as ResolveExternalsConfig);
  defaultExternals.forEach(a => alias.addAlias(a));

  const runConfig = loadConfiguration(process.cwd(), configPath);

  const userWebpackConfig =
    runConfig?.config &&
    (runConfig.config.webpack || runConfig.config.finalWebpack);

  if (!userWebpackConfig) {
    return webpackConfig;
  }
  return typeof userWebpackConfig === 'function'
    ? userWebpackConfig(webpackConfig, options)
    : deepMergeWebpackConfig(webpackConfig, userWebpackConfig);
};

export const runCompiler = (
  run: (
    compiler: Compiler,
    callback: (err: Error, stats: Stats) => void,
  ) => void,

  props: CompileRunProps,
): Promise<CompileResults> => {
  return new Promise((resolve, reject) => {
    //@ts-ignore
    const compiler = webpack(createConfig(props));
    run(compiler, (err, stats) => {
      if (err) {
        console.error(err);
        return reject(err);
      }
      if (stats.hasErrors() || stats.hasWarnings()) {
        const error = stats.toString({
          errorDetails: true,
          warnings: true,
        });
        console.error(error);
        return reject(error);
      }

      const { store } = require('@component-controls/loader/store');
      console.log(
        chalk.bgRgb(244, 147, 66)(`compiled ${store.stores.length} documents`),
        require.resolve('@component-controls/loader/store'),
      );
      resolve({ store, stats });
    });
  });
};
