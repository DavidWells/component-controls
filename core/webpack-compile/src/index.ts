const chalk = require('chalk');
import {
  CompileProps,
  CompileResults,
  WatchProps,
} from '@component-controls/webpack-configs';
import { runCompiler, CompilerCallbackFn } from './utilities';

export { CompilerCallbackFn };

/**
 * compile the stories with webpack
 * returns the stories store object
 */
export const compile = (
  {
    webPack,
    presets,
    configPath,
    staticFolder,
    distFolder,
    bundleName,
  }: CompileProps,
  callback?: CompilerCallbackFn,
): Promise<CompileResults> => {
  console.log(
    chalk.bgRgb(244, 147, 66)('@start compilation'),
    'optimized build',
  );
  return runCompiler(
    (compiler, callback) => compiler.run(callback),
    {
      webPack,
      mode: 'production',
      presets,
      configPath,
      staticFolder,
      distFolder,
      bundleName,
    },
    callback,
  );
};

/**
 * compile the stories with webpack and launch watching for changes
 * returns the stories store object
 */
export const watch = (
  {
    webPack,
    presets,
    configPath,
    watchOptions,
    staticFolder,
    distFolder,
    bundleName,
  }: WatchProps,
  callback?: CompilerCallbackFn,
): Promise<CompileResults> => {
  console.log(
    chalk.bgRgb(244, 147, 66)('@start compilation'),
    'development mode watch',
  );
  return runCompiler(
    (compiler, callback) => compiler.watch({ ...watchOptions }, callback),
    {
      webPack,
      mode: 'development',
      presets,
      configPath,
      staticFolder,
      distFolder,
      bundleName,
    },
    callback,
  );
};
