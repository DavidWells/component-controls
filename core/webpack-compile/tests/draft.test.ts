import * as path from 'path';
import { compile } from '../src/index';

expect.addSnapshotSerializer({
  test: val => typeof val === 'object' && val instanceof String,
  print: val => {
    return `"${val.toString().trim()}"`;
  },
});

describe('test draft', () => {
  it('compile', async () => {
    const { bundleName } = await compile({
      presets: ['react', 'react-docgen-typescript'],
      configPath: path.resolve(__dirname, 'fixtures', 'draft'),
    });
    const bundle = require(bundleName);
    expect(bundle).toMatchSnapshot();
  }, 50000);
});
