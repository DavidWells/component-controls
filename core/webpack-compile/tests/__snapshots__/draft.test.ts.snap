// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test draft compile 1`] = `
Object {
  "buildConfig": Object {
    "categories": Array [
      "author",
      "tags",
    ],
    "pages": Object {
      "author": Object {
        "basePath": "authors/",
      },
      "blog": Object {
        "basePath": "blogs/",
      },
      "page": Object {
        "basePath": "pages/",
      },
      "story": Object {
        "basePath": "docs/",
        "sideNav": Object {
          "collapseSingle": true,
          "storyPaths": true,
        },
        "tabs": Array [
          Object {
            "route": "page",
          },
          Object {
            "route": "test",
          },
        ],
      },
      "tags": Object {
        "basePath": "tags/",
      },
    },
    "stories": Array [
      "./*.stories.(js|jsx|tsx|mdx)",
    ],
  },
  "components": Object {},
  "config": undefined,
  "packages": Object {},
  "stores": Array [],
}
`;
